src/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -479,19 +480,26 @@ public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompl
             CancellationTokenSource cts;
             bool disposeCts;
             bool hasTimeout = _timeout != s_infiniteTimeout;
+            CancellationTokenSource pendingRequestsCts = _pendingRequestsCts;
+            Func<Exception, bool> isTimeout = NoTimeoutPredicate;
             if (hasTimeout || cancellationToken.CanBeCanceled)
             {
                 disposeCts = true;
-                cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _pendingRequestsCts.Token);
+                cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, pendingRequestsCts.Token);
                 if (hasTimeout)
                 {
+                    isTimeout = e =>
+                        e is OperationCanceledException
+                        && cts.IsCancellationRequested
+                        && !pendingRequestsCts.IsCancellationRequested
+                        && !cancellationToken.IsCancellationRequested;
```

Comment:
This is the ugly part. The check will actually be made in other methods (`FinishSendAsyncUnbuffered` or `FinishSendAsyncBuffered`), so I need to pass these methods what they need to check if the exception is a timeout. That means I have to pass either:
- `hasTimeout`, `pendingRequestsCts` and `cancellationToken`: I dislike this approach because `FinishSendAsyncUnbuffered` and `FinishSendAsyncBuffered` aren't supposed to know the details of how we create the CTS and its relationship with `_pendingRequestsCts` ad `cancellationToken`
- a delegate that encapsulate that logic (what I've done here). It's better, but it means more allocations (for the closure and for the delegate), which I'd rather avoid.

If requested, I can try another approach: creating a struct that encapsulates the logic and pass it to the `FinishSendAsync*` methods. This would avoid the allocation, but it would be a relatively big struct.

Reply:
> I dislike this approach because FinishSendAsyncUnbuffered and FinishSendAsyncBuffered aren't supposed to know the details of how we create the CTS and its relationship with _pendingRequestsCts ad cancellationToken

I see your point but this still concerns implementation detail, so I think we can tolerate a tiny bit of ugliness if it means saving a few redundant allocations and virtual calls. Can go a long way improving with careful choice of naming and comments explaining stuff.

Reply:
OK, I'll change it then.

BTW, there's another approach that would make things simpler: introduce a new `CancellationTokenSource` for the timeout. It would make it easy to check if a timeout has really occurred. However it implies an extra allocation...

Reply:
I suspect introducing more layers of linked CTS's would entail more complexity than just one allocation.

Reply:
> If requested, I can try another approach: creating a struct that encapsulates the logic and pass it to the `FinishSendAsync*` methods. This would avoid the allocation, but it would be a relatively big struct.

I pushed a commit implementing this. The struct isn't that big in fact (just two fields, a CT and a CTS), and it keeps the code clean. If you don't like it, I can unwrap it, but it means passing the `FinishSendAsync*` methods arguments whose purpose are not obvious.